### PR TITLE
fix(react-streamdown): update streamed code block content

### PR DESCRIPTION
**Pull request by Kinfe123**
## Summary

- compare nested React-element props in `memoCompareNodes` instead of treating matching type/key as equal
- refresh only the memoized fenced-code child when streamed content is replaced
- preserve the existing code component instance during prefix-only token appends

## Why

`PreOverride` previously skipped updates when its child kept the same type and key even though its text props changed. Comparing nested props makes that update observable, but Streamdown 2.5's nested code memo can still retain the old subtree when the HAST source position is unchanged.

The fix keeps a small transactional revision at the fenced-code boundary. The revision stays stable for ordinary append-only streaming such as `fir` to `first`, and changes for a replacement such as `first` to `second`. This refreshes the stale code child without remounting the surrounding markdown tree or remounting code on every token.

## Validation

- reproduced the default-renderer failure on untouched `origin/main`
- added nested React-element comparator coverage
- added default-renderer regressions for streaming, static, and deferred replacements
- added a mount-count regression proving prefix appends preserve the code component instance
- `pnpm --filter @assistant-ui/react-streamdown test` (143 tests)
- `pnpm --filter @assistant-ui/react-streamdown build`
- targeted oxlint and oxfmt checks

<a href="https://rupic.dev/s/s2.ncdnMOjXtYWqgl3l1GF_2sO8QLzbtgHXyJx6OpuztEY8k2RfFPZvU7u0Mcg?ref=github" target="_blank" rel="noopener noreferrer"><picture><source media="(prefers-color-scheme: dark)" srcset="https://rupic.dev/badges/track-dark-v1.svg">Track in Rupic</picture></a>

### CODE REVIEW

File: packages/react-streamdown/src/__tests__/StreamdownText.test.tsx [hunk 1]
```diff
@@ -91,6 +91,25 @@ describe("StreamdownTextPrimitive", () => {
...
+    rerender(
+      <TextMessagePartProvider text={"```ts\nsecond\n```"} isRunning>
+        <StreamdownTextPrimitive />
+      </TextMessagePartProvider>,
+    );
+
+    expect(await screen.findByText("second")).toBeTruthy();
+    expect(screen.queryByText("first")).toBeNull();
+  });
+
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-streamdown/src/__tests__/StreamdownText.test.tsx]: Confirmed the fenced-code replacement test is red on origin/main and green with the fix. f42dd2832 also adds mount-counting tests for normalizeMathDelimiters across an append boundary and for the defer path.

File: packages/react-streamdown/src/primitives/StreamdownText.tsx [hunk 2]
```diff
@@ -174,6 +175,11 @@ export const StreamdownTextPrimitive = forwardRef<
 
     const deferredText = useDeferredValue(text);
     const processedText = defer ? deferredText : text;
+    const streamRevisionRef = useRef({ text: processedText, revision: 0 });
+    if (!processedText.startsWith(streamRevisionRef.current.text)) {
+      streamRevisionRef.current.revision += 1;
+    }
+    streamRevisionRef.current.text = processedText;
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-streamdown/src/primitives/StreamdownText.tsx]: Addressed in f42dd2832. Discontinuity detection now uses raw runtime text before preprocess, smooth, and defer, and revision history uses React state rather than a render-mutated ref. Added no-remount coverage for a math preprocessing rewrite and deferred appends.

File: packages/react-streamdown/src/primitives/StreamdownText.tsx [hunk 3]
```diff
@@ -174,6 +175,11 @@ export const StreamdownTextPrimitive = forwardRef<
 
     const deferredText = useDeferredValue(text);
     const processedText = defer ? deferredText : text;
+    const streamRevisionRef = useRef({ text: processedText, revision: 0 });
+    if (!processedText.startsWith(streamRevisionRef.current.text)) {
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-streamdown/src/primitives/StreamdownText.tsx]: Addressed in 8b6cb8a2c. The revision now also tracks the subscribed aui.part identity, matching the established useSmooth pattern, so a true part replacement resets Streamdown even when its text preserves the old prefix.

File: packages/react-streamdown/src/primitives/StreamdownText.tsx [hunk 4]
```diff
@@ -29,6 +30,17 @@ import type {
 
 type StreamdownTextPrimitiveElement = ComponentRef<"div">;
 
+const useStreamRevision = (text: string) => {
+  const [state, setState] = useState({ text, revision: 0 });
+  if (state.text === text) return state.revision;
+
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-streamdown/src/primitives/StreamdownText.tsx]: Addressed in 8b6cb8a2c. Stream revision state now includes the subscribed part identity and increments whenever that identity changes, while retaining the raw-text prefix fast path for appends within the same part.

File: packages/react-streamdown/src/__tests__/StreamdownText.test.tsx [hunk 5]
```diff
@@ -91,6 +92,83 @@ describe("StreamdownTextPrimitive", () => {
...
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-streamdown/src/__tests__/StreamdownText.test.tsx:111]: Added in 8b6cb8a2c. The regression resets a real LocalRuntime message from first to the prefix-preserving first second, uses a stateful block to expose stale content, and asserts the replacement remounts and removes the stale rendering.

File: packages/react-streamdown/src/__tests__/StreamdownText.test.tsx [hunk 6]
```diff
@@ -91,6 +100,124 @@ describe("StreamdownTextPrimitive", () => {
...
+  it("resets blocks for prefix-preserving message part replacements", async () => {
+    const mounted = vi.fn();
+    const BlockComponent = ({ content }: { content: string }) => {
+      const [renderedContent] = useState(content);
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-streamdown/src/__tests__/StreamdownText.test.tsx]: Agreed on the synthetic test. It has been deleted. The retained regression uses the default Streamdown renderer and fails on untouched main with first still visible after rerendering second. The custom BlockComponent is now used only in a separate performance guard proving that prefix appends do not remount.

File: packages/react-streamdown/src/primitives/StreamdownText.tsx [hunk 7]
```diff
@@ -29,6 +35,18 @@ import type {
 
 type StreamdownTextPrimitiveElement = ComponentRef<"div">;
 
+const useStreamRevision = (text: string, part: unknown) => {
+  const [state, setState] = useState({ text, part, revision: 0 });
+  if (state.text === text && state.part === part) return state.revision;
+
+  const revision =
+    state.part === part && text.startsWith(state.text)
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-streamdown/src/primitives/StreamdownText.tsx]: The implementation no longer reads aui.part or relies on its facade identity. Revision changes are content-based and use the processed text actually sent toward Streamdown. Prefix-preserving changes keep the subtree because the upstream node position changes and rerenders correctly; only non-prefix replacements need the workaround for equal-position stale code.

File: packages/react-streamdown/src/__tests__/StreamdownText.test.tsx [hunk 8]
```diff
@@ -91,6 +92,83 @@ describe("StreamdownTextPrimitive", () => {
...
+          BlockComponent={BlockComponent}
+        />
+      </TextMessagePartProvider>,
+    );
+
+    expect(await screen.findByText("$x$")).toBeTruthy();
+    expect(mounted).toHaveBeenCalledTimes(1);
+  });
+
+  it("does not remount blocks for deferred text appends", async () => {
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-streamdown/src/__tests__/StreamdownText.test.tsx]: Agreed. I removed the discarded-render claim and the old happy-path test. The current deferred regression has a narrower purpose: it verifies that replacement detection follows processedText after useDeferredValue, so the key changes when the deferred replacement is actually passed to Streamdown.

File: packages/react-streamdown/src/primitives/StreamdownText.tsx [hunk 9]
```diff
@@ -257,6 +270,7 @@ export const StreamdownTextPrimitive = forwardRef<
         className={containerClass}
       >
         <Streamdown
+          key={streamRevision}
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-streamdown/src/primitives/StreamdownText.tsx]: Confirmed in a detached worktree at base 9fc00788 with only the default-renderer regression test added: Vitest failed with 1 failed test, Unable to find an element with the text: second, while the DOM still contained first. The stale layer is Streamdown 2.5 MarkdownCode, whose comparator checks className and HAST source positions; equal-length replacements retain those positions. The revision now follows processedText, no longer tracks part identity, and preserves prefix appends. I also reproduced the same failure with explicit static mode, so the PR now covers that real path rather than describing this as running-only.

File: packages/react-streamdown/src/__tests__/StreamdownText.test.tsx [hunk 10]
```diff
@@ -91,6 +100,124 @@ describe("StreamdownTextPrimitive", () => {
...
+          <ThreadPrimitive.Messages components={{ Message }} />
+        </AssistantRuntimeProvider>
+      );
+    };
+
+    render(<App />);
+    expect(await screen.findByText("first")).toBeTruthy();
+
+    await act(async () => {
+      runtime!.thread.reset([{ role: "assistant", content: "first second" }]);
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-streamdown/src/__tests__/StreamdownText.test.tsx]: The thread.reset test and its generated-message-ID ambiguity have been removed. The regression now rerenders the default renderer in place under the same TextMessagePartProvider location, and it fails on untouched main without any outer runtime replacement.

File: packages/react-streamdown/src/primitives/StreamdownText.tsx [hunk 11]
```diff
@@ -160,7 +178,10 @@ export const StreamdownTextPrimitive = forwardRef<
     },
     ref,
   ) => {
+    const aui = useAui();
+    const part = useAuiState(() => aui.part);
     const messagePart = useMessagePartText();
+    const streamRevision = useStreamRevision(messagePart.text, part);
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-streamdown/src/primitives/StreamdownText.tsx]: Fixed in 3a443001d. useStreamRevision now receives processedText after useDeferredValue rather than messagePart.text. A default-renderer deferred replacement regression verifies that second replaces first.

File: packages/react-streamdown/src/__tests__/StreamdownText.test.tsx [hunk 12]
```diff
@@ -91,6 +100,124 @@ describe("StreamdownTextPrimitive", () => {
...
+          BlockComponent={BlockComponent}
+        />
+      </TextMessagePartProvider>,
+    );
+
+    expect(await screen.findByText("$x$")).toBeTruthy();
+    expect(mounted).toHaveBeenCalledTimes(1);
+  });
+
+  it("does not remount blocks for deferred text appends", async () => {
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-streamdown/src/__tests__/StreamdownText.test.tsx]: Agreed. The PR no longer claims to test a discarded background render. The current test is explicitly deferred replacement coverage for key and displayed-text alignment.

File: packages/react-streamdown/src/primitives/StreamdownText.tsx [hunk 13]
```diff
@@ -257,6 +278,7 @@ export const StreamdownTextPrimitive = forwardRef<
         className={containerClass}
       >
         <Streamdown
+          key={streamRevision}
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-streamdown/src/primitives/StreamdownText.tsx]: I reproduced explicit static mode failing on untouched main with the default renderer, so this is not limited to running streams. Streamdown 2.5 reuses equal-position MarkdownCode nodes in both paths. The implementation now avoids part-identity resets and preserves prefix appends, while non-prefix static replacements reset the stale subtree.

File: packages/react-streamdown/src/primitives/StreamdownText.tsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-streamdown/src/primitives/StreamdownText.tsx]: The part-identity behavior has been removed. I also tested explicit mode static against untouched main and it has the same user-visible failure: second is missing and first remains. The key now changes only for a non-prefix processed-text replacement; prefix updates preserve the subtree. Static replacement remains covered because Streamdown 2.5 has the same stale source-position memo there.

File: packages/react-streamdown/src/primitives/StreamdownText.tsx [hunk 14]
```diff
@@ -29,6 +30,19 @@ import type {
 
 type StreamdownTextPrimitiveElement = ComponentRef<"div">;
 
+// Streamdown 2.5 memoizes code nodes by source position, so equal-length
+// replacements need a new subtree while ordinary prefix appends stay mounted.
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-streamdown/src/primitives/StreamdownText.tsx]: Verified against the installed Streamdown 2.5 bundle and an untouched-main reproduction. The parent `Block` does rerender, but its child `MarkdownCode` is memoized by `className` plus HAST source position; replacing fenced content can retain the fence positions and therefore the stale child subtree. The default renderer reproduced `first` remaining after `second` in both streaming and explicit static modes. Commit 525705961 updates the comment and PR body to name that component and removes the incorrect equal-length claim.

File: packages/react-streamdown/src/primitives/StreamdownText.tsx [hunk 15]
```diff
@@ -174,6 +188,7 @@ export const StreamdownTextPrimitive = forwardRef<
 
     const deferredText = useDeferredValue(text);
     const processedText = defer ? deferredText : text;
+    const streamRevision = useStreamRevision(processedText);
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-streamdown/src/primitives/StreamdownText.tsx]: Valid concern, fixed in 525705961. Replacement classification now uses raw `messagePart.text`, while rendered text is tracked only so a deferred replacement resets when the new rendered value is ready. The preprocess and deferred prefix-append mount-count checks are restored.

File: packages/react-streamdown/src/primitives/StreamdownText.tsx [hunk 16]
```diff
@@ -257,6 +272,7 @@ export const StreamdownTextPrimitive = forwardRef<
         className={containerClass}
       >
         <Streamdown
+          key={streamRevision}
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-streamdown/src/primitives/StreamdownText.tsx]: I reproduced the same stale default-renderer output on untouched main with explicit `mode="static"`, so the defect is not limited to running messages. The PR description now states that scope. The key still changes only for non-prefix replacements, so ordinary settled rendering and prefix appends retain the existing tree.

File: packages/react-streamdown/src/primitives/StreamdownText.tsx (hunk 15 above)
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-streamdown/src/primitives/StreamdownText.tsx]: Fixed in 525705961. Raw `messagePart.text` now decides whether the update is a replacement; processed/rendered text only controls when a delayed replacement reset is applied. This preserves normal preprocess, smoothing, and deferred prefix appends.

File: packages/react-streamdown/src/__tests__/StreamdownText.test.tsx [hunk 17]
```diff
@@ -91,6 +91,64 @@ describe("StreamdownTextPrimitive", () => {
...
+          <StreamdownTextPrimitive {...props} />
+        </TextMessagePartProvider>,
+      );
+
+      expect(await screen.findByText("second")).toBeTruthy();
+      expect(screen.queryByText("first")).toBeNull();
+    },
+  );
+
+  it("preserves streaming blocks for prefix appends", async () => {
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-streamdown/src/__tests__/StreamdownText.test.tsx]: Fixed in 525705961. The mount-counting regression is now table-driven across plain, `normalizeMathDelimiters`, and deferred prefix appends, and all three assert a single mount.

File: packages/react-streamdown/src/primitives/StreamdownText.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-streamdown/src/primitives/StreamdownText.tsx]: Updated in 525705961 after tracing the installed Streamdown 2.5 bundle. The parent `Block` rerenders, but the nested `MarkdownCode` memo compares `className` and HAST source position, which can retain stale fenced-code content. The comment now names that specific mechanism.

File: packages/react-streamdown/src/primitives/StreamdownText.tsx [hunk 18]
```diff
@@ -29,6 +30,31 @@ import type {
...
+    renderedText,
+    pendingReset: false,
+    revision: 0,
+  });
+  if (state.rawText === rawText && state.renderedText === renderedText) {
+    return state.revision;
+  }
+
+  let pendingReset =
+    state.pendingReset ||
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-streamdown/src/primitives/StreamdownText.tsx]: Addressed in 5cbeda60b by tracking the canonical `aui.part` identity, the same signal `useSmooth` uses for part/thread switches. A new part now resets even when its text extends the previous prefix; a prefix update on the same part remains an ordinary append.

File: packages/react-streamdown/src/primitives/StreamdownText.tsx [hunk 19]
```diff
@@ -257,6 +284,7 @@ export const StreamdownTextPrimitive = forwardRef<
         className={containerClass}
       >
         <Streamdown
+          key={streamRevision}
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-streamdown/src/primitives/StreamdownText.tsx]: I verified this scope directly: the default renderer remains stale on untouched main with explicit `mode="static"` after replacing the fenced content. Static/settled replacements therefore need the same reset. The key does not change for ordinary appends, and the PR body now calls out the static reproduction explicitly.

File: packages/react-streamdown/src/primitives/StreamdownText.tsx [hunk 20]
```diff
@@ -29,6 +30,31 @@ import type {
 
 type StreamdownTextPrimitiveElement = ComponentRef<"div">;
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-streamdown/src/primitives/StreamdownText.tsx:31]: Updated in 5cbeda60b. The code comment now states only the verified invariant: replaced rendered text needs a fresh subtree, while prefix appends preserve incremental state. The detailed upstream trace remains in the PR description and is backed by the untouched-main reproduction.

File: packages/react-streamdown/src/primitives/StreamdownText.tsx [hunk 21]
```diff
@@ -29,6 +30,31 @@ import type {
...
+  let pendingReset =
+    state.pendingReset ||
+    (state.rawText !== rawText && !rawText.startsWith(state.rawText));
+  let revision = state.revision;
+  if (pendingReset && state.renderedText !== renderedText) {
+    pendingReset = false;
+    revision += 1;
+  }
+  setState({ rawText, renderedText, pendingReset, revision });
+  return revision;
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-streamdown/src/primitives/StreamdownText.tsx]: Valid performance concern, fixed in 5cbeda60b. Normal appends now update refs only and never call `setState`; state changes only when a replacement is actually rendered. The pending flag is also cleared when a subsequent raw append supersedes a replacement that rendered to the same output. Canonical `aui.part` identity handles prefix-shaped part replacements without sacrificing the normal append path.

File: packages/react-streamdown/src/primitives/StreamdownText.tsx [hunk 22]
```diff
@@ -29,6 +36,44 @@ import type {
...
+  const previous = previousRef.current;
+  const rawTextChanged = rawText !== previous.rawText;
+  const textReplaced =
+    part !== previous.part ||
+    (rawTextChanged && !rawText.startsWith(previous.rawText));
+
+  if (textReplaced) {
+    pendingResetRef.current = true;
+  } else if (rawTextChanged) {
+    pendingResetRef.current = false;
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-streamdown/src/primitives/StreamdownText.tsx]: Resolved by replacing the revision-key workaround in 503152c9d. `useStreamRevision` no longer exists, so deferred appends cannot clear a pending reset. The fix now refreshes only the memoized fenced-code child when its actual content changes.

File: packages/react-streamdown/src/primitives/StreamdownText.tsx [hunk 23]
```diff
@@ -29,6 +36,44 @@ import type {
...
 
+// Replaced text needs a fresh subtree once the replacement is rendered, while
+// prefix appends preserve Streamdown's incremental state.
+const useStreamRevision = (
+  part: unknown,
+  rawText: string,
+  renderedText: string,
+) => {
+  const [revision, setRevision] = useState(0);
+  const previousRef = useRef({
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-streamdown/src/primitives/StreamdownText.tsx]: Resolved by removing part-identity replacement detection entirely in 503152c9d. The fix is now content-local: `memoCompareNodes` observes changed child props and `PreOverride` keys only the fenced-code child by its primitive content.

File: packages/react-streamdown/src/__tests__/StreamdownText.test.tsx [hunk 24]
```diff
@@ -91,6 +92,91 @@ describe("StreamdownTextPrimitive", () => {
...
+        <TextMessagePartProvider
+          text={"```ts\nfirst\n```"}
+          isRunning={isRunning}
+        >
+          <StreamdownTextPrimitive {...props} />
+        </TextMessagePartProvider>,
+      );
+
+      expect(await screen.findByText("first")).toBeTruthy();
+
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-streamdown/src/__tests__/StreamdownText.test.tsx:111]: The replacement-generation workaround has been removed. The regression now exercises the actual failing boundary with the default renderer: the same mounted primitive changes fenced content from `first` to `second`, and the child-prop comparator/content key makes that update observable without relying on part identity.

File: packages/react-streamdown/src/primitives/StreamdownText.tsx [hunk 25]
```diff
@@ -257,6 +309,7 @@ export const StreamdownTextPrimitive = forwardRef<
         className={containerClass}
       >
         <Streamdown
+          key={streamRevision}
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-streamdown/src/primitives/StreamdownText.tsx]: Resolved in 503152c9d. There is no longer an unconditional key on `<Streamdown>`, so static or settled replacements do not remount the full markdown tree. Only the cloned fenced-code child receives a content key.

File: packages/react-streamdown/src/primitives/StreamdownText.tsx [hunk 26]
```diff
@@ -29,6 +36,44 @@ import type {
...
+    (rawTextChanged && !rawText.startsWith(previous.rawText));
+
+  if (textReplaced) {
+    pendingResetRef.current = true;
+  } else if (rawTextChanged) {
+    pendingResetRef.current = false;
+  }
+
+  const shouldReset =
+    pendingResetRef.current && renderedText !== previous.renderedText;
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-streamdown/src/primitives/StreamdownText.tsx]: Resolved by deleting the latch and revision hook in 503152c9d. The targeted comparator plus fenced-child key has no deferred pending state to lose, and the deferred default-renderer regression passes.

File: packages/react-streamdown/src/primitives/StreamdownText.tsx [hunk 27]
```diff
@@ -29,6 +36,44 @@ import type {
...
+  renderedText: string,
+) => {
+  const [revision, setRevision] = useState(0);
+  const previousRef = useRef({
+    part,
+    rawText,
+    renderedText,
+  });
+  const pendingResetRef = useRef(false);
+  const previous = previousRef.current;
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-streamdown/src/primitives/StreamdownText.tsx]: Resolved by deleting all render-phase refs/state from `StreamdownTextPrimitive` in 503152c9d. The fix now lives at the memoized fenced-code boundary, so interrupted deferred renders cannot corrupt replacement bookkeeping.

File: packages/react-streamdown/src/adapters/PreOverride.tsx [hunk 28]
```diff
@@ -47,9 +47,17 @@ export const PreOverride = memo(function PreOverride({
...
+    "data-block"?: string;
+    children?: ReactNode;
+  }>(children)
+    ? cloneElement(children, {
         "data-block": "true",
+        key:
+          typeof children.props.children === "string" ||
+          typeof children.props.children === "number"
+            ? children.props.children
+            : children.key,
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-streamdown/src/adapters/PreOverride.tsx]: Confirmed the comparator-only version still failed all three default-renderer regressions because Streamdown 2.5 also memoizes the nested code subtree by stable source position. Addressed in 5bc23279d: the fenced-code child now uses a state-backed revision that stays stable for prefix appends and increments only for a replacement. The new mount-count test proves fir -> first remains one mount while first -> second remounts once; all 141 package tests pass.

File: packages/react-streamdown/src/adapters/PreOverride.tsx [hunk 29]
```diff
@@ -47,9 +47,17 @@ export const PreOverride = memo(function PreOverride({
...
-  const childWithBlock = isValidElement(children)
-    ? cloneElement(children as ReactElement<{ "data-block"?: string }>, {
+  const childWithBlock = isValidElement<{
+    "data-block"?: string;
+    children?: ReactNode;
+  }>(children)
+    ? cloneElement(children, {
         "data-block": "true",
+        key:
+          typeof children.props.children === "string" ||
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-streamdown/src/adapters/PreOverride.tsx]: Addressed in 5bc23279d. The key is no longer derived from every content value: it uses a transactional revision that remains stable for prefix appends and changes only on replacement. A focused mount-count test verifies streamed appends retain the existing code component instance.

File: packages/react-streamdown/src/adapters/PreOverride.tsx [hunk 30]
```diff
@@ -47,9 +47,17 @@ export const PreOverride = memo(function PreOverride({
...
+          typeof children.props.children === "number"
+            ? children.props.children
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-streamdown/src/adapters/PreOverride.tsx]: Addressed in 5bc23279d. The key is now based on a transactional revision, not the text itself: fir -> first retains revision 0 and the same mount, while first -> second advances the revision once. The focused mount-count regression covers this exact append path.

File: packages/react-streamdown/src/memoization.ts [hunk 31]
```diff
@@ -16,7 +26,9 @@ function isReactElement(node: unknown): node is ReactElement {
 function compareNodes(a: ReactNode, b: ReactNode): boolean {
   if (a === b) return true;
   if (!isReactElement(a) || !isReactElement(b)) return false;
-  return a.type === b.type && a.key === b.key;
+  return (
+    a.type === b.type && a.key === b.key && memoCompareNodes(a.props, b.props)
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-streamdown/src/memoization.ts:36]: Addressed in e488dfc29. React-node arrays are now compared by length and recursively element-by-element, with a regression covering equivalent newly allocated child arrays.

File: packages/react-streamdown/src/adapters/PreOverride.tsx (hunk 29 above)
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-streamdown/src/adapters/PreOverride.tsx]: Addressed in 5bc23279d. Appended code content keeps the same revision key; only a non-prefix replacement increments it. The regression asserts a single mount across fir -> first and a single remount for first -> second.

File: packages/react-streamdown/src/__tests__/StreamdownText.test.tsx [hunk 32]
```diff
@@ -91,6 +91,38 @@ describe("StreamdownTextPrimitive", () => {
     expect(await screen.findByText("first chunk second chunk")).toBeTruthy();
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-streamdown/src/__tests__/StreamdownText.test.tsx:91]: An equivalent mount-count guard now lives in PreOverride.test.tsx at the exact keyed boundary introduced by this fix. It asserts prefix streaming keeps one mount and replacement causes one remount. Preprocess and deferred values no longer participate in any Streamdown-level revision bookkeeping, because that earlier full-tree mechanism was removed.

File: packages/react-streamdown/src/adapters/PreOverride.tsx [hunk 33]
```diff
@@ -3,19 +3,51 @@
...
+    content === state.anchor ||
+    (state.anchor !== null && content.startsWith(state.anchor))
+  ) {
+    return state.revision;
+  }
+
+  if (state.anchor === null || state.anchor === "") {
+    setState({ anchor: content, revision: state.revision });
+    return state.revision;
+  }
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-streamdown/src/adapters/PreOverride.tsx]: Addressed in 7abc058ab. Accepted prefix appends now update the transactional anchor without changing the revision, so later replacements compare against the latest streamed content rather than the initial fragment. The existing fir -> first -> second mount test covers that sequence.

File: packages/react-streamdown/src/adapters/PreOverride.tsx [hunk 34]
```diff
@@ -3,19 +3,51 @@
...
+    revision: 0,
+  }));
+
+  if (
+    content === null ||
+    content === state.anchor ||
+    (state.anchor !== null && content.startsWith(state.anchor))
+  ) {
+    return state.revision;
+  }
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-streamdown/src/adapters/PreOverride.tsx]: Addressed in 7abc058ab. The renderer-added trailing newline is removed before prefix comparison, the anchor advances on every accepted append, and the unreachable empty-anchor branch is gone. A new mount-count regression runs through StreamdownTextPrimitive with real fenced markdown and proves fir -> first remains one code-component mount. The package now passes all 143 tests.

File: packages/react-streamdown/src/__tests__/PreOverride.test.tsx [hunk 35]
```diff
@@ -153,6 +153,36 @@ describe("PreOverride component", () => {
...
+        <Code>fir</Code>
+      </PreOverride>,
+    );
+
+    rerender(
+      <PreOverride>
+        <Code>first</Code>
+      </PreOverride>,
+    );
+    expect(onMount).toHaveBeenCalledTimes(1);
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-streamdown/src/__tests__/PreOverride.test.tsx:175]: Addressed in 7abc058ab. The direct fixture now includes the renderer's trailing newline, and a second mount-count regression streams fenced markdown through the real StreamdownTextPrimitive pipeline and asserts one mount across fir -> first.

File: packages/react-streamdown/src/adapters/PreOverride.tsx [hunk 36]
```diff
@@ -3,19 +3,47 @@
...
   node?: Element | undefined;
 };
 
+const useCodeRevision = (children: ReactNode) => {
+  const childContent = isValidElement<{ children?: ReactNode }>(children)
+    ? children.props.children
+    : null;
+  const content =
+    typeof childContent === "string" || typeof childContent === "number"
+      ? String(childContent).replace(/\n$/, "")
```
[the remark this answers is not in the document]
Comment (Kinfe123) [reply packages/react-streamdown/src/adapters/PreOverride.tsx:26]: Confirmed. This is separate from the stale memoization bug: preprocess currently treats an unterminated fence body as prose, so its transformed code content can legitimately become non-prefix before the fence closes. Fixing that belongs in preprocess/codeSpanEnd with its own focused regression rather than expanding this PR.

**Comment by okisdev** [comment]
superseded by #6330. streamdown 2.6 already compares children in its code memo, so the in-repo revision counter is not needed.

**Comment by Kinfe123** [comment]
Oh nice didn't know it was live.
